lib/recursive-methods.rb

```diff
@@ -1,49 +1,94 @@
+require "pry"
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n recursive calls are made
+# Space complexity: O(n) where the stack is n levels deep
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError if n < 0
+  if (n == 1) || (n == 0)
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n recursive calls are made
+# Space complexity: O(n) where n levels are required, each requiring 1 char less of memory
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 1 || s.length == 0
+    return s
+  else
+    return s[-1] + reverse(s[0...-1])
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n/2 recursive calls are made
+# Space complexity: O(n) where the max depth of the function is n/2
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 1 || s.length == 0
+    return s
+  else
+    s[-1] + reverse_inplace(s[1...-1]) + s[0]
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n recursive calls are made
+# Space complexity: O(n) where the max depth is n
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  if n <= 0
+    return 0
+  else
+    return 2 + bunny(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n/2 recursive calls are made
+# Space complexity: O(n) where the max depth of the function is n/2
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 0
+    return true
+  elsif s.length == 1
+    return false
+  elsif s[0] == s[-1]
+    return false
+  else
+    nested(s[1...s.length - 1])
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) because there are 2n-1 calls each time.
+# Space complexity: O(log n) because at most the algorithm goes log2 n levels deep
 def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+  return false if array.length == 0
+  if array.length == 1
+    return true if array[0] == value
+  else
+    first_half = array[0...(array.length / 2)]
```

Comment:
This looks like an attempt at MergeSort or something.  Because you're making new arrays, this will be O(n^2), but I don't see what advantage you're getting by splitting it. 

This could be easier as:

```ruby
return false if array.length == 0
return true if value == array[0]
return search(array[1...-1])
```

Reply:
Wouldn't this save stack memory (space complexity) though? At most, the method will only go log2(n) levels deep at a time, as opposed to n levels if you just knocked off one number with each call.

Reply:
It will save on some stack space, but won't save you in time complexity.